microhapulator/api.py

```diff
@@ -687,24 +687,25 @@ def plot_haplotype_calls(result, outdir, sample=None, plot_marker_name=True, ign
 def get_reads_in_marker_loci(fullref_bam_file, mhindex):
     fullref_bam = pysam.AlignmentFile(fullref_bam_file, "rb")
     reads_to_markers = defaultdict(list)
-    for locus in mhindex.loci.values():
-        start, end = locus.span(chrom=True)
+    for locus, marker in mhindex:
+        start, end = marker.span(chrom=True)
         for read in fullref_bam.fetch(locus.chrom, start, end):
             if not skip_read(read):
-                reads_to_markers[read.query_name].append(locus.id)
+                reads_to_markers[read.query_name].append(marker.id)
     return reads_to_markers
 
 
-def count_repetitive_reads(marker_bam, locus, reads_to_markers, minbasequal):
+def count_repetitive_reads(marker_bam, locus, marker, reads_to_markers_fullref, minbasequal):
     repetitive_count = 0
     for read in marker_bam.fetch(locus.id):
         if skip_read(read):
             continue
-        qual_mask = np.array(read.query_qualities) >= minbasequal
-        qual_filt_positions = np.array(read.get_reference_positions(full_length=True))[qual_mask]
-        contains_varloc = bool(set(qual_filt_positions) & locus.offsets())
-        if contains_varloc:
-            repetitive_count += locus.id not in reads_to_markers[read.query_name]
+        ref_pos = read.get_reference_positions(full_length=True)
+        quals = read.query_qualities
+        qual_filtered_pos = [pos for i, pos in enumerate(ref_pos) if quals[i] >= minbasequal]
+        qual_snp_pos = [offset for offset in marker.offsets_locus if offset in qual_filtered_pos]
+        if qual_snp_pos and marker.id not in reads_to_markers_fullref[read.query_name]:
+            repetitive_count += 1
     return repetitive_count
```

Comment:
This was the function with the weird boolean logic. As far as I can tell it was behaving as intended, but I made some changes to (hopefully) make it more clear.

Reply:
It is indeed more clear, but we might be able to do even better. I think the `zip` function iterating over two lists is a bit cleaner than the enumerate to index into the second list.

```python
refr_pos = read.get_reference_positions(full_length=True)
quals = read.query_qualities
qual_filtered_pos = [pos for pos, qual in zip(refr_pos, quals) if qual >= minbasequal]
```

More importantly, the conditional statement on line 707 is still pretty inscrutable.

- The first half, `if qual_snp_pos`, is concise but hides what is actually being tested. That variable is a list: what do you mean by `if list`? I'm pretty sure casing a list as a boolean in Python will return `False` for an empty list and `True` otherwise. If that's the intended behavior here (as I suspect), the code should be more explicit: `if len(qual_snp_pos) > 0`. Or we could store the result in a boolean variable with a descriptive name: `overlaps_target_snps = len(qual_snp_pos) > 0; if overlaps_target_snps and...`.
- The purpose of the second half of the conditional is also unclear at first glance. This is a check to see if the read maps elsewhere in the genome, right? So maybe something like `maps_elsewhere = marker.id not in ...`.

So all together, my suggestion would be something like this.

```diff
     for read in marker_bam.fetch(locus.id):
         if skip_read(read):
             continue
-        ref_pos = read.get_reference_positions(full_length=True)
+        refr_pos = read.get_reference_positions(full_length=True)
         quals = read.query_qualities
-        qual_filtered_pos = [pos for i, pos in enumerate(ref_pos) if quals[i] >= minbasequal]
+        qual_filtered_pos = [pos for pos, qual in zip(refr_pos, quals) if qual >= minbasequal]
         qual_snp_pos = [offset for offset in marker.offsets_locus if offset in qual_filtered_pos]
-        if qual_snp_pos and marker.id not in reads_to_markers_fullref[read.query_name]:
+        read_overlaps_target_snps = len(qual_snp_pos) > 0
+        read_maps_elsewhere = marker.id not in reads_to_markers_fullref[read.query_name]
+        if read_overlaps_target_snps and read_maps_elsewhere:
             repetitive_count += 1
     return repetitive_count
```

I'm happy to discuss more if I've misunderstood something, or if you want to discuss other possible variable names or strategies.